Experiment: Convert gradle to Kotlin

Rover656 opened the pull request:
# Description

I have converted our build scripts to use Kotlin DSL instead of Groovy - mainly due to the improved structure it promises to provide.

This might be done naively, so feel free to review critically :)

<!-- For drafts, fill this in as you go; if you are leaving draft, make sure these are all done -->
# Checklist:

- [x] My code follows the style guidelines of this project (.editorconfig, most IDEs will use this for you).
- [x] I have performed a self-review of my own code.
- [x] I have commented my code in areas it may be challenging to understand. <!-- (Although we prefer code that is readable instead of over-commented) -->
- [x] I have made corresponding changes to the documentation.
- [x] My changes are ready for review from a contributor.

<!-- Thanks to: https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/ for the building blocks of this template -->

Rover656 committed:
Experiment: Convert build scripts to Kotlin DSL.

Rover656 committed:
Upgrade NeoForge to support Mekanism.

justliliandev commented:
do you really expect us to review gradle stuff? I have no idea what's going on there

Rover656 commented:
> do you really expect us to review gradle stuff? I have no idea what's going on there

The expected response 😂 I'll do another review pass and merge it tomorrow :p

hypherionmc commented:
I could also take a look over it. I have some understanding of kts build scripts.

Just a note I want to add, modpublisher 2.0+ has much better kotlin support than 1.0. The docs and github have been updated with the new changes

Rover656 commented:
> I could also take a look over it. I have some understanding of kts build scripts.
> 
> Just a note I want to add, modpublisher 2.0+ has much better kotlin support than 1.0. The docs and github have been updated with the new changes

That'd be wonderful if you have the time! I'll also take a look at upgrading modpublisher later tonight :)

ferriarnus commented on gradle.properties:
```diff
@@ -25,10 +25,11 @@ jei_version=17.3.0.49
 #top_version=5.0.0
 top_version=1.20.0-10.0.0
 patchouli_version=1.19.3-78
-mekanism_cf_id=3922056
```
I would use the mekanism maven, see my Chemical conduits PR.

Rover656 replied:
Good call, I'll change it to use the mek maven :)

Rover656 committed:
Switch to Mekanism maven and address jarJar TODO for graphLib

Rover656 committed:
Upgrade mod publisher to 2.0

Rover656 committed:
Fixed Implementation-Timestamp in manifest

Rover656 commented:
That should be all current feedback addressed :)

Rover656 committed:
Return to classic versioning scheme. More in description.

During the release of v6.0.9-alpha, the Minecraft version was added to the version string. This quickly changed however to appending it to the jar name instead, as it was breaking some versioning systems behind the scenes.
This workaround is no longer required now that we're launching a new release series.

Rover656 committed:
Revert "Return to classic versioning scheme. More in description."

This reverts commit 5a933d6590cda8092815ceb38f74ac674128b7f9.

Rover656 committed:
Working on changing up our versioning scheme further

Rover656 committed:
Show minecraft versions in Discord webhooks

Rover656 committed:
Upload datagen changes

Rover656 committed:
Stick a 1 in front of dev versions to make it appear valid when running datagen :P
